include/unifex/detail/debug_async_scope.hpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/detail/intrusive_list.hpp>
+
+#include <cxxabi.h>
+#include <mutex>
+#include <typeinfo>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex::detail {
+namespace _debug_async_scope {
+
+struct op_base {
+  op_base(const std::type_info& concreteType) noexcept
+    : concreteType(concreteType)
+    , demangled(demangle(concreteType)) {}
+
+  const std::type_info& concreteType;
+  const char* demangled;
+
+  static const char* demangle(const std::type_info& concreteType) noexcept {
+    int status = -1;
+    auto result = abi::__cxa_demangle(concreteType.name(), nullptr, 0, &status);
+    UNIFEX_ASSERT(status == 0);
+    return result;
```

Comment:
I feel like this might be better if it checked `status` and did something like `return "";` or `return "<failed to demangle>";` or even `return concreteType.name();` on failure.

Also, where does `result` point when this succeeds?  Do you need to deallocate something?

Reply:
From https://gcc.gnu.org/onlinedocs/libstdc++/manual/ext_demangling.html

> The demangler interface is described in the source documentation linked to above. It is actually written in C, so you don't need to be writing C++ in order to demangle C++. (That also means we have to use crummy memory management facilities, so don't forget to free() the returned char array.)

Looks like we need to call `free()`.